packages/structures/src/Structure.ts

```diff
@@ -0,0 +1,133 @@
+/* eslint-disable jsdoc/check-param-names */
+import { kData, kMixinConstruct } from './utils/symbols.js';
+import type { ReplaceOmittedWithUnknown } from './utils/types.js';
+
+/**
+ * Additional options needed to appropriately construct / patch data
+ *
+ * @internal
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface StructureExtraOptions {}
+
+export const DataTemplatePropertyName = 'DataTemplate';
+export const OptimizeDataPropertyName = '_optimizeData';
+
+/**
+ * Represents a data model from the Discord API
+ *
+ * @privateRemarks
+ * Explanation of the type complexity surround Structure:
+ *
+ * There are two layers of Omitted generics, one here, which allows omitting things at the library level so we do not accidentally
+ * access them, in addition to whatever the user does at the layer above.
+ *
+ * The second layer, in the exported structure is effectively a type cast that allows the getters types to match whatever data template is used
+ *
+ * In order to safely set and access this data, the constructor and patch take data as "partial" and forcibly assigns it to kData. To acommodate this,
+ * kData stores properties as `unknown` when it is omitted, which allows accessing the property in getters even when it may not actually be present.
+ * This is the most technically correct way of representing the value, especially since there is no way to guarantee runtime matches the "type cast."
+ */
+export abstract class Structure<DataType, Omitted extends keyof DataType | '' = ''> {
+	/**
+	 * A construct function used when mixing to allow mixins to set optimized property defaults
+	 *
+	 * @internal
+	 * @remarks This should only be used to set defaults, setting optimized values should be done
+	 * in the mixins `_optimizeData` method, which will be called automatically.
+	 * @param data - The full API data received by the Structure
+	 */
+	protected [kMixinConstruct]?(data: Partial<DataType>): void;
+
+	/**
+	 * The template used for removing data from the raw data stored for each Structure.
+	 *
+	 * @remarks This template should be overriden in all subclasses to provide more accurate type information.
+	 * The template in the base {@link Structure} class will have no effect on most subclasses for this reason.
+	 */
+	protected static DataTemplate: Record<string, unknown> = {};
```

Comment:
perhaps I'm wrong, but I don't see us mutating this anywhere

```suggestion
	protected readonly static DataTemplate: Record<string, unknown> = {};
```

or sorry, I forget which one our linter likes, might be
```suggestion
	protected static readonly DataTemplate: Record<string, unknown> = {};
```